app/services/billable_metrics/aggregations/unique_count_service.rb

```diff
@@ -67,36 +67,39 @@ def compute_per_event_aggregation
 
       protected
 
-      # This method fetches the latest event in current period. If such a event exists we know that metadata
-      # with previous aggregation and previous maximum aggregation are stored there. Fetching these metadata values
+      # This method fetches the latest cached aggregation in current period. If such a record exists we know that
+      # previous aggregation and previous maximum aggregation are stored there. Fetching these values
       # would help us in pay in advance value calculation without iterating through all events in current period
-      def previous_event
-        @previous_event ||= begin
-          query = if billable_metric.recurring?
-            recurring_events_scope(to_datetime:, from_datetime:)
-          else
-            events_scope(from_datetime:, to_datetime:)
-          end
-          query = query
-            .joins(:quantified_event)
-            .where(field_presence_condition)
-            .where("events.metadata->>'current_aggregation' IS NOT NULL")
-            .where("events.metadata->>'max_aggregation' IS NOT NULL")
-            .where('quantified_events.added_at::timestamp(0) >= ?', from_datetime)
-            .where('quantified_events.added_at::timestamp(0) <= ?', to_datetime)
-
-          query = query.where.not(id: event.id) if event.present?
-          query = query.reorder(created_at: :desc)
-
-          query
-            .where('quantified_events.removed_at::timestamp(0) IS NULL')
-            .or(
-              query
-                .where('quantified_events.removed_at::timestamp(0) >= ?', from_datetime)
-                .where('quantified_events.removed_at::timestamp(0) <= ?', to_datetime),
-            )
-            .first
-        end
+      def cached_aggregation
+        return @cached_aggregation if @cached_aggregation
+
+        query = CachedAggregation
+          .where(organization_id: billable_metric.organization_id)
+          .where(external_subscription_id: subscription.external_id)
+          .where(charge_id: charge.id)
+          .from_datetime(from_datetime)
+          .to_datetime(to_datetime)
+          .order(created_at: :desc)
```

Comment:
for sum aggregation we order by timestamp. WDYT? Should we stick to timestamp?

Reply:
You are right, it makes senses to use timestamp :+1: